### PR TITLE
including several comments from WG members

### PR DESCRIPTION
**Pull request by ksriram25**
@job @waehlisch @tanneberger @mitradir
As promised to the Chairs and authors in the email thread last week, I have worked over the last two weeks to make changes based on comments from several people in the SIDROPS WG since v-16. This includes:

1. Based on a suggestion from Tassilo, added a section on "Semantics of Valid, Invalid, and Unknown". 
2. Also based on a suggestion from Tassilo, added a section on "Basic Principles of Route Leak Detection and Reason for Separate Upstream and Downstream Algorithms".
3. Based on suggestions from Rudiger and Doug, expanded the Security Considerations section to include more examples.
4. Incorporated comments received from Alexander in September 2023. 
5. Refined the section on "Properties of ASPA-based Path Verification" and added an additional property.
6. Refined the "Terminology and List of Acronyms" section that was added by TT/MW.
7. Refined the "Implementation and Deployment Recommendations" section that was added by TT/MW.

Overall, I think this version should put us across the finish line with the WGLC.
- Sriram

### CODE REVIEW

**Comment by mitradir** [comment]
I will merge it after merging 'mitradir' branch. It would be great if you apply your changes to it.

**Comment by ksriram25** [comment]
@mitradir @job @tanneberger 

> I will merge it after merging 'mitradir' branch. It would be great if you apply your changes to it.

Alexander: Sorry to say this but please do not merge the mitradir branch currently. We should first publish the PR#22 as version-17 (of course, including any comments received). Your changes created in the 'mitradir' branch cause a large churn. Not clear why it is necessary. Let us be aware that about 30 WG members actively participated, and they have systematically worked through and contributed to versions 12 through 16 in the WGLC period between March to August 2023. They did all this based on comments, discussion, and consensus. There is broad consensus now on the algorithm descriptions and the draft overall.   

Four implementers have understood the draft very well and have already implemented it. They have expressed no technical issues or concerns with the draft. Even Tassilo had no problem implementing after he spoke with me in Prague and got answers to his questions. He is not suggesting that the algorithms lack correctness or need any major revision. I have included his comments to date already in the PR#22. Even you are not changing the algorithms, you just seem to prefer a different way of using the indices. That is OK to do in your own implementation.

Also, there were discussions and convergence on other details in the draft which you seem to be changing. For example, the mutual-transit case -- the definition and handling of it was discussed and agreed on. Each AS is a transit provider of the other AS for all prefixes. We used to call this sibling, but Jay Borkenhagen suggested it should be called mutual transit. He and the WG were good with including it in the draft. Then the algorithms correctly accounted for it. 

You also mentioned something about reconsidering AFI in your edits. What do you have on mind? I hope you are aware that the WG has discussed it at length and has strong consensus to make ASPA AFI agnostic. 

If you wish to propose the major overhaul of the draft, it would be best to first post some messages on the WG list to describe the changes and let people understand why you think such overhaul is needed. Then everyone would have a chance to react and comment. Thanks.

**Comment by mitradir** [comment]
Sriram, there was feedback that verification procedures, the key part of the document, it is very hard to read and become a block for implementors. To goal is to create a technology, not a document, right?

You are very welcome to comment if I deleted something important in 'mitradir' branch. But I tried to do my best to keep all good that was added by you and WG in the document during last year. If there are issues, let's discuss it case by case.

For the AFI flag, as you may noticed, we already had several discussions about AFI flag. We made it optional, and implementors tried to follow this change. When it proved to be hard we removed it for good, without precise study on the consequences. At least once we already wasted valuable resources of our fellow implementors. If you have time to measure the effect of removing AFI flag, it would be great.

**Comment by cjeker** [comment]
> For the AFI flag, as you may noticed, we already had several discussions about AFI flag. We made it optional, and implementors tried to follow this change. When it proved to be hard we removed it for good, without precise study on the consequences. At least once we already wasted valuable resources of our fellow implementors. If you have time to measure the effect of removing AFI flag, it would be great.

There is close to no valid study you can do since you don't know if a provider would create an AFI specific ASPA record in the first place. In most cases adding AFI specific ASPA is more of a foot gun then preventing some leak.
If you can't trust your provider for one AFI but not for the other you should probably rethink your choice.

The complication that AFI specific ASPA records added to a) the spec, b) the operators publishing ASPA, c) the operators trying to figure out why a path turned invalid and d) the implementors having to juggle even more options, just made this button unattractive and more of a burden. Why do we need to constantly come back to this?

**Comment by mitradir** [comment]
> > For the AFI flag, as you may noticed, we already had several discussions about AFI flag. We made it optional, and implementors tried to follow this change. When it proved to be hard we removed it for good, without precise study on the consequences. At least once we already wasted valuable resources of our fellow implementors. If you have time to measure the effect of removing AFI flag, it would be great.
> 
> There is close to no valid study you can do since you don't know if a provider would create an AFI specific ASPA record in the first place. In most cases adding AFI specific ASPA is more of a foot gun then preventing some leak. If you can't trust your provider for one AFI but not for the other you should probably rethink your choice.
> 
> The complication that AFI specific ASPA records added to a) the spec, b) the operators publishing ASPA, c) the operators trying to figure out why a path turned invalid and d) the implementors having to juggle even more options, just made this button unattractive and more of a burden. Why do we need to constantly come back to this?

I see your points. But my question is not about trust, it is about policy and measurements if IPv4 and IPv6 are really on a steady convergence process.

Anyway, in the candidate 'mitradir' branch, I haven't reintroduced the AFI flag.

**Comment by tanneberger** [comment]
> Four implementers have understood the draft very well and have already implemented it. They have expressed no technical 
> issues or concerns with the draft. Even Tassilo had no problem implementing after he spoke with me in Prague and got answers > to his questions.

@ksriram25 Correctly implementing the draft and understanding it without much prior knowledge took about as long as the entire IETF week + hackathon. In my opinion, our goal should not be a correct formulation of the algorithm, but additionally an understandable one. It is much better if the implementer understands what he is writing.

**Comment by ksriram25** [comment]
@tanneberger @cjeker  @waehlisch 
> Correctly implementing the draft and understanding it without much prior knowledge took about as long as the entire IETF week + hackathon. 

Wow! All of this accomplished in 7 days while attending the IETF. That is superfast. I am honestly very impressed. 

You (@tanneberger) earlier gave these comments (in email):
>2.) Add a section on the semantics of VALID, INVALID and UNKNOWN
>3.) Revise the Algorithm Section and add an improve introductory section explaining why it is necessary to consider downstream and upstream validation separately. ...

I have followed up on that and I have included two new sub-sections in PR#22 as follows:
    7.1.  Semantics of Valid, Invalid, and Unknown
    7.2.  Basic Principles of Route Leak Detection and Reason for
           Separate Upstream and Downstream Algorithms

Have you taken a look at those sections? The new Section 7.2 explains the basic principles of AS path verification with a clear diagram. Once the reader reads these two sections, then following the algorithms (Sec. 7.3 and 7.4) should become quite easy.

Sriram

**Comment by mitradir** [comment]
Sriram, to code basic algo with the document at hand for the experienced developer should take hours, not days with people standing around and providing guidance.

I'm nearly ready to push my branch to master. @ksriram25 I would ask you to make a rebase and merge latest changes and after that send an updated patch.

**Comment by cjeker** [comment]
> > Four implementers have understood the draft very well and have already implemented it. They have expressed no technical issues or concerns with the draft. Even Tassilo had no problem implementing after he spoke with me in Prague and got answers  to his questions.
> 
> @ksriram25 Correctly implementing the draft and understanding it without much prior knowledge took about as long as the entire IETF week + hackathon. In my opinion, our goal should not be a correct formulation of the algorithm, but additionally an understandable one. It is much better if the implementer understands what he is writing.

I agree with this. When I implemented ASPA validation I had to use the slides from a presentation to better understand what and how it is validated. The RFC text is too abstract to follow and we already improved the situation.
The algorithm section reads like a science paper and it is not very approachable unless you already have the right mental image of what happens.

It also did not help that what I implemented in the end is quite different from what is written in the text. The behaviour is the same but many checks are done differently.

File: draft-ietf-sidrops-aspa-verification.xml [hunk 1]
```diff
@@ -541,92 +602,117 @@ hop(AS(i), AS(j)) =  / Else, "Provider+" if the VAP-SPAS entry
...
+     If the outcome of applying the upstream algorithm (<xref target="Upflow"/>) is Invalid, then the route SHOULD NOT be propagated to a transit provider, lateral peer, RS (local AS has RS-client role), or RS-client (local AS has RS role).
+     If the outcome of applying the downstream algorithm (<xref target="Downflow"/>) is Invalid, then the route SHOULD NOT be propagated to an eBGP neighbor regardless of its BGP role (<xref target="role"/>).
+   </t>
+  </section>
+  -->
+   <section title="Only to Customer (OTC) Attribute" anchor="otc">
+   <t>
+    While the ASPA-based AS_PATH verification method (<xref target="verif"/>, <xref target="impl-verif"/>) detects and mitigates route leaks that were
+    created by preceding ASes listed in the AS_PATH, it lacks the ability to prevent route leaks from occurring at the local AS.
+    The use of the Only to Customer (OTC) Attribute <xref target="RFC9234"/> fills in that gap.
```
Comment (cjeker) [inline draft-ietf-sidrops-aspa-verification.xml:669]: I understand the idea of telling people that `OTC` and `ASPA` are orthogonal and both should be used,
but I don't think that OTC prevents route leaks at the local AS (route leaks are always between ASs).

File: draft-ietf-sidrops-aspa-verification.xml [hunk 2]
```diff
@@ -737,36 +824,100 @@ hop(AS(i), AS(j)) =  / Else, "Provider+" if the VAP-SPAS entry
...
+ASPAs: {AS(1), [AS(2)]}, {AS(2), [AS(3)]}, {AS(3), [AS 0]},
+       {AS(4), AS(3)}, {AS(5), AS(4)}, {AS(6), [AS(5)]}
+]]>
+        </artwork>
+      </figure>
+    </t>
+    <t>
+      In the above scenario, it may also be noted that if the attacking AS forwards the data traffic destined
+      for the affected prefix, and if all other ASes physically connected to the attacking AS are ASPA compliant,
+      then the data traffic (from the deceived customer) would flow on a feasible route-leak free path in spite of the attack.
```
Comment (cjeker) [inline draft-ietf-sidrops-aspa-verification.xml:877]: I don't think there is a need to explain which path manipulations can and which can't be detected by ASPA.
The detection of path manipulations is more of a side-effect of what ASPA provides but I would not sell ASPA as secure.
It makes it harder which is great but it feels out of scope for this document.

There is a lot of text added about various kind of hijacks. I don't think ASPA really provides a major hurdle here -- it helps and raises the bar for sure but like ROA it does not prevent hijacks.
Especially sophisticated attacks done by cyber criminals and hacker groups.

**Review by cjeker — COMMENTED** [review]
See inline comments.
A lot of the changes are minor but I think Section 7.2 does not improve the explanation of the algorithms. Also a lot of text was added to explain something that is IMO out of scope. ASPA is not here to prevent sophisticated hijacks. It makes it harder but it does not prevent them.

**Comment by ksriram25** [comment]
@mitradir 
> I'm nearly ready to push my branch to master. @ksriram25 I would ask you to make a rebase and merge latest changes and after that send an updated patch.

Alexander: There appears to be a significant overhaul in your branch. Do you want us to look at it and try to have some pre-merge comments/discussion?

**Comment by mitradir** [comment]
Sure, if you see any issues in the specification, please let me know.
I would suggest you wait until I merge it into the master, so you can apply your edits. I'm going to merge it until the end of the week.

File: draft-ietf-sidrops-aspa-verification.xml [hunk 3]
```diff
@@ -335,20 +332,84 @@ hop(AS(i), AS(j)) =  / Else, "Provider+" if the VAP-SPAS entry
         In the current document, routes with AS_SET are given Invalid evaluation in the AS_PATH verification procedures (<xref target="Upflow"/> and <xref target="Downflow"/>).
         See <xref target="mitigation"/> for how routes with Invalid AS_PATH are handled.
       </t>
+
+    <section title="Semantics of Valid, Invalid, and Unknown" anchor="semantics">
+      <t>
+        The ASPA-based path verification algorithms (<xref target="Upflow"/> and <xref target="Downflow"/>) evaluate an AS_PATH as Valid, Invalid, or Unknown.
+        'Invalid' outcome means that the AS_PATH contains a route leak (i.e., valley-free violation) based on the available ASPAs.
+        Even if the actual attack on the AS_PATH attribute was a forged-origin hijack or forged-path-segment hijack
```
[the remark this answers is not in the document]
Comment (ksriram25) [reply draft-ietf-sidrops-aspa-verification.xml:340]: Agree, will remove the particular sentence.

File: draft-ietf-sidrops-aspa-verification.xml [hunk 4]
```diff
@@ -162,48 +162,45 @@
...
                 </t>
-
                 <t hangText="VAP:">
-		   A Validated ASPA Payload, see <xref target="ASPA"/>.
+                  Validated ASPA Payload (<xref target="ASPA"/>).
                 </t>
-
                 <t hangText="VAP-SPAS:">
-		   A Validated ASPA Payload of a set of provider ASes, see <xref target="ASPA"/>.
+                  Set of Provider ASes contained in Validated ASPA Payload (<xref target="ASPA"/>).
```
[the remark this answers is not in the document]
Comment (ksriram25) [reply draft-ietf-sidrops-aspa-verification.xml:196]: VAP-SPAS are indeed validated ASPA payloads. Yes, non-validated payloads are eliminated and only VAP-SPAS remain for use in ASPA path verification. Analogy: If you think back about ROA and ROV, the VRP (Validated ROA Payload) is used in RFC 6811 for ROV. VAP-SPAS is analogous to VRP.

File: draft-ietf-sidrops-aspa-verification.xml [hunk 5]
```diff
@@ -162,48 +162,45 @@
...
+                </t>
+                <t hangText="Provider+:">
+                  The term "Provider+" in the definition of the hop-check function is meant to encompass all three possibilities: Provider, non-transparent RS, or mutual-transit neighbor (<xref target="pair-validation"/>).
```
[the remark this answers is not in the document]
Comment (ksriram25) [reply draft-ietf-sidrops-aspa-verification.xml:199]: This is based on WGLC feedback. During WGLC, Amir Herzberg emphasized that use of Provider+ is essential. Otherwise, it gets confusing to simply use the term Provider here to represent all flavors such as Provider, RS, and mutual-transit neighbor.

File: draft-ietf-sidrops-aspa-verification.xml [hunk 6]
```diff
@@ -335,20 +332,84 @@ hop(AS(i), AS(j)) =  / Else, "Provider+" if the VAP-SPAS entry
...
+      </t>
       <t>
-        In <xref target="Upflow"/> and <xref target="Downflow"/> below, the terms "upstream path" and "downstream path" generally refer to AS paths received in the upstream direction (from a customer or a lateral peer) and in the downstream direction (from a provider or a mutual-transit neighbor), respectively.
-        An RS-client receiving a route from its RS is a special case where the algorithm for upstream paths is applied (<xref target="Upflow"/>).
+        Let some acronyms be defined for AS hop types as follows: P2C = provider-to-customer, C2P = customer-to-provider, and P2P = peer-to-peer (i.e., lateral peers).
+        Consider the full AS path of a BGP Update consisting of sequential AS hops from the origin AS to the receiving/verifying AS.
+        Let the unique ASes in this path be represented by AS(k), i = 1, 2,... , N+1, where AS(1) is the origin AS and AS(N+1) is the receiving/verifying AS.
+        As illustrated in <xref target="fig5"/>, the AS path is considered a route leak (i.e., valley free violation <xref target="RFC7908"/>) if for
+        some i and j (j &ge; i+1), the hop AS(i) to AS(i+1) is either a P2C or p2p while
```
[the remark this answers is not in the document]
Comment (ksriram25) [reply draft-ietf-sidrops-aspa-verification.xml:357]: Thank for catching that. Will fix it.

File: draft-ietf-sidrops-aspa-verification.xml [hunk 7]
```diff
@@ -335,20 +332,84 @@ hop(AS(i), AS(j)) =  / Else, "Provider+" if the VAP-SPAS entry
...
+        Even if the actual attack on the AS_PATH attribute was a forged-origin hijack or forged-path-segment hijack
+        or other AS path modification,
+        and the associated conditions permit its detectability by ASPA (see <xref target="property"/>), it would be effectively manifest as a route leak and hence ASPA would detect the route as Invalid.
+        The algorithms do not further identify the specific type of anomaly.
+        'Valid' outcome means that sufficient ASPA information is available to determine that the AS_PATH attribute as presented is route-leak free (i.e., valley free). When the outcome is not Valid or Invalid, then it is Unknown by default.
+      </t>
+    </section>
+    <section title="Basic Principles of Route Leak Detection and Reason for Separate Upstream and Downstream Algorithms" anchor="basic">
+      <t>
+        In <xref target="Upflow"/> and <xref target="Downflow"/> below, the terms "upstream path" and "downstream path" generally refer to AS paths received in the upstream direction (from a customer or a lateral peer) and the downstream direction (from a provider), respectively.
```
[the remark this answers is not in the document]
Comment (ksriram25) [reply draft-ietf-sidrops-aspa-verification.xml:349]: Thanks for the catch. They are just typos. 
s/ received in the upstream direction (from a customer or a lateral peer) and the downstream direction (from a provider) / received in the 'up' direction (i.e., from a customer or a lateral peer) and the 'down' direction (i.e., from a provider) /

File: draft-ietf-sidrops-aspa-verification.xml [hunk 8]
```diff
@@ -335,20 +332,84 @@ hop(AS(i), AS(j)) =  / Else, "Provider+" if the VAP-SPAS entry
...
+                           \/           
+                          A(j)
+
+ (D) .... AS(j+1)---(P2P)--->A(j) ..... AS(i+1)<---P2P---AS(i) ....
+                               
+]]>
+</artwork>
+        </figure>
       </t>
+    </section>
```
[the remark this answers is not in the document]
Comment (ksriram25) [reply draft-ietf-sidrops-aspa-verification.xml:402]: I think you are referring to Section 7.2. It can be improved as follows. The text should upfront refer to RFC7908 where the four key types of route leaks (valley-free violations) are defined in Secs 3.1 through 3.4. Basically, a route leak happens if the update is sent from a provider to a customer or between lateral peers on one hop and subsequently sent from a customer to a provider or between lateral peers on another hop. The two hops where this happens need not be adjacent.

File: draft-ietf-sidrops-aspa-verification.xml (hunk 1 above)
Comment (ksriram25) [reply draft-ietf-sidrops-aspa-verification.xml:669]: >I don't think that OTC prevents route leaks at the local AS (route leaks are always between ASs).

Good catch. The wording is not right. We didn't mean a route leak internal to the AS. 
s/ it lacks the ability to prevent route leaks from occurring at the local AS / it lacks the ability to prevent the local AS from initiating a route leak towards its neighbor /
OTC is used as tag from the ingress to the egress and is used on egress to prevent the local AS from being a leaker. RFC 9234 wording is as follows: "In the case of prevention at the local AS, the presence of an OTC Attribute indicates to the egress router that the route was learned from a Peer, a Provider, or an RS, and it can be advertised only to the Customers."

File: draft-ietf-sidrops-aspa-verification.xml (hunk 2 above)
Comment (ksriram25) [reply draft-ietf-sidrops-aspa-verification.xml:877]: I was indeed in doubt about including the specific paragraph that you highlighted. Will remove it. 

>There is a lot of text added about various kind of hijacks. I don't think ASPA really provides a major hurdle here -- it helps and raises the bar for sure but like ROA it does not prevent hijacks.

What you are stating is not entirely correct. We have known it all along and WG members certainly appreciated that ASPA does prevent forged-origin and forged-path-segment hijacks in the *upstream paths*. But the same is not true for the downstream paths. Folks understood and liked the *upstream paths* part and they appreciated the text about that during WGLC discussions.  

WGLC participants also liked highlighting an example of undetected AS path manipulation in the Security Considerations section. Rudiger strongly advised including multiple examples because he thought only one example was inadequate and misleading. There are in indeed at least three distinct types that are undetected by ASPA, and I have included them in the Security Considerations section.  

Thank you, Claudio, for the comments. Very helpful

**Comment by cjeker** [comment]
> > > 2.) Add a section on the semantics of VALID, INVALID and UNKNOWN
> > > 3.) Revise the Algorithm Section and add an improve introductory section explaining why it is necessary to consider downstream and upstream validation separately. ...
> > 
> > 
> > I have followed up on that and I have included two new sub-sections in PR#22 as follows: 7.1. Semantics of Valid, Invalid, and Unknown 7.2. Basic Principles of Route Leak Detection and Reason for Separate Upstream and Downstream Algorithms
> > Have you taken a look at those sections? The new Section 7.2 explains the basic principles of AS path verification with a clear diagram. Once the reader reads these two sections, then following the algorithms (Sec. 7.3 and 7.4) should become quite easy.
> 
> Algorithm-Section: Aside from the textual problems with this text, I think the main point I was trying to make that is missing from this text is the fact that when you start implementing this, you start thinking about why these two separate algorithms (up vs. down validation) are not combined. And then the normal developer would try to combine the algorithms and end up failing, or worse, have a very subtle, divergent implementation (with potentially very serious negative implications). We have basically gone through all the stages of despair to understand that it's not really worth combining the two algorithms. I would like to see an explanation as to why there is this weird split into two algorithms, and perhaps point to our attempts. So I think we need to take another jab at it.

When implementing this in OpenBGPD I did not have this issue. Now the implementation in OpenBGPD is calculating
both upflow and downflow together (the hop-check function is called in both directions for each AS pair) but I would not call this combining the algorithms. What I agree with is that it took a lot of effort to understand what ASPA does and why it does it the way it is done. Maybe it helped to have implemented RFC9234 `OTC` first (I knew about roles already) since I never questioned the split of the algorithm. What I had more issues with is what you mention below.

> Semantics: This is particularly tricky because the algorithm lives in a very incomplete and very abstracted world and it is very easy to get lost in it. The current draft and forget how to translate it back into the real world. The reader should get a sense of what certain ASPA annotations mean in the context of real-world network deployment.

I agree that the algorithm is very abstracted and introduces a lot of new (and somewhat unnecessary) terminology and works with directions (up, down, first, last) that are not always clear. Compared to previous version the current draft already improved a lot but there is still potential for improvements but overall the problem is that this is a complex beast with lots of conditions to check.

File: draft-ietf-sidrops-aspa-verification.xml (hunk 4 above)
[the remark this answers is not in the document]
Comment (cjeker) [reply draft-ietf-sidrops-aspa-verification.xml:196]: In my opinion a `VAP` is analogous to `VRP`? It is 'validated ASPA payload' and 'validated ROA payload'.
The fact that there are `SPAS` (a set of provider ASes) in the ASPA payload is not that relevant.
Nobody introduced a  `VRP-OAS` (validation roa payload origin AS) terminology in the ROA specifications.
See the comment from @tanneberger further below, the ASPA draft introduces a lot of terminology which makes it easy for the reader to get lost.

Isn't it enough to mention in section 3 that only validated ASPA records should be used in the validation?
How the validation is done is described in `I-D.ietf-sidrops-aspa-profile` and how the data is loaded into the router is out of scope (either via RTRv2 or some other means (like local config)).
In other words a router only needs `VAP` entries and these `VAP`s contain a `set of provider ASes:  SPAS`
There is no need to be so verbose.

File: draft-ietf-sidrops-aspa-verification.xml (hunk 5 above)
[the remark this answers is not in the document]
Comment (cjeker) [reply draft-ietf-sidrops-aspa-verification.xml:199]: I think I mentioned it before, the hop-check function should stay away from such terminology.
It would be better to describe the results of the hop-check function as:
- found, not-found, unknown or no-attestation
- in-set, not-in-set, unknown or no-attestation
- or some other form that is not connected to the term provider or AS

Hop-check is a simple lookup function with 3 possible results that checks if `provider_as` is an attested provider of `customer_as` using the present VAP data.

```
hop_check(customer_as, provider_as)
```
a) if there is no entry for `customer-as` in the `VAP` database return `no-attestation`
b) if there is an entry for `provider-as` in the `SPAS` of the `customer_as` `VAP` return `found`
c) else return `not-found`

Why do we need to introduce yet another terminology and abstraction around `Provider+` that needs to be explained? I think in Section 5 about half of the text should be removed.

File: draft-ietf-sidrops-aspa-verification.xml (hunk 7 above)
[the remark this answers is not in the document]
Comment (cjeker) [reply draft-ietf-sidrops-aspa-verification.xml:349]: While better I would prefer something where `upflow` is not explained with the term `up`.
The problem is that `up` and `down` make sense in e.g. a building but it is far less clear in this context.
My other issue is that roles are also depending on the view point. When I configure a neighbor on a router do I specify his role towards me or is it my role towards him. This little difference suddenly requires to rethink `up` and `down` since the roles are swapped.

I know this is a tough nut to crack but it is one of the main issues I had to fight with during implementation.
Mapping `up`, `down`, `first`, `last`, `customer`, `provider`, ... to my environment and therefor my mental image.
This is why I requested to write and number the as sequences in a form that people expect.
Every time a direction term is used it adds load to the mental image people need to create.

File: draft-ietf-sidrops-aspa-verification.xml (hunk 1 above)
Comment (cjeker) [reply draft-ietf-sidrops-aspa-verification.xml:669]: > s/ it lacks the ability to prevent route leaks from occurring at the local AS / it lacks the ability to prevent the local AS from initiating a route leak towards its neighbor /

I think this change helps. Even OTC does not detect all route leaks initiated by the local AS (even if the provider AS has RFC9234 implemented) since it depends on the transitive nature of the `OTC` attribute. But this is out of scope here.

The goal is to tell everyone to implement both :)

**Comment by mitradir** [comment]
@ksriram25 I merged my branch with the master, please intercorporate your edits on top of it. Or you can send me edits so I can help you with merging.

File: draft-ietf-sidrops-aspa-verification.xml (hunk 2 above)
Comment (cjeker) [reply draft-ietf-sidrops-aspa-verification.xml:877]: I do understand that ASPA prevents some form of forged hijacks but in my opinion it gives a false sense of security.

It only works in some cases (_upstream paths_) and only if all possible providers reached in the graph from the origin-AS up to tier-1 AS use ASPA. Right now a hijack can be done via unknown hops (provider with no ASPA record published) which is sadly the most common case right now and for the foreseeable future.

It is a great side-effect and one of the reasons I really push for ASPA to become widespread but I think the bar for the Security Considerations section should be higher.

**Comment by ksriram25** [comment]
@mitradir 
>>>Alexander: There appears to be a significant overhaul in your branch. Do you want us to look at it and try to have some pre-merge comments/discussion?

>>Sure, if you see any issues in the specification, please let me know.
I would suggest you wait until I merge it into the master, so you can apply your edits. I'm going to merge it until the end of the week.

>I merged my branch with the master, please intercorporate your edits on top of it. Or you can send me edits so I can help you with merging.

Alexander: We had the above conversation and you said you were going to wait. Then you merged it quickly anyway. Why the rush? The rest of us are following a discipline. We first create a PR and let everyone think and comment on it. We do the merge when there is consensus after the discussions settle down. Can you also please adhere to this discipline also?  

We need to have patience and listen to other's comments and discussion. We have a responsibility towards the SIDROPS WG members who gave a lot of inputs and they have overwhelmingly supported the document to pass the WGLC. Please look back at the SIDROPS list archives. We cannot throw away their inputs at this stage. It would be disrespectful and doing a disservice to the WG and the IETF. 

If you can undo or put aside the merge and still create a PR with your changes, please do. Let us give it chance to be discussed carefully as a PR.

I will personally do the combing with my changes after discussion settles down on your changes. I am receiving energetic discussion/feedback on my PR#22 as you are seeing. I will take all that into consideration carefully when I update the document. 

Since I have been the lead and point person in the WGLC discussions over the last year handing several revisions (v-13 through v-16), please allow me to finalize the next update v-17 for IETF upload. I need to make sure all inputs are carefully addressed, and no WG member is offended as best as possible. 

It all works out very well if we listen to each other and work hard to achieve a common goal. I can say this from my experience with multiple RFCs that are enjoying success in some form or shape.

**Comment by mitradir** [comment]
Sriram, it would be much easire and useful to discuss your PR after rebase.

I have had several discussions after I shared the link with co-authors and other contributors. The changes that were suggested were applied. You were invited to work on it too.

My update was based on the latest version of master branch, so it should cover all improvements during the WGLC. If you find any issues don't hesitate to email me or you may try to use issue button at the repo.

**Comment by ksriram25** [comment]
@mitradir 
>Sriram, it would be much easire and useful to discuss your PR after rebase.

There has been good feedback and discussion on my PR that I will carefully take into consideration. Yes, I'll plan to rebase. I need time to review your changes carefully. Will get back. Thanks for your patience.

**Comment by ksriram25** [comment]
@mitradir @job 
Just a heads up. I have rebased to latest Master and working on edits/updates. It is shaping up well. It will take a few more days to complete and issue a new PR. The comments received on this PR#22 are carefully being taken into consideration. Please also see the comments I wrote for Alexander in the Master branch:
https://github.com/QratorLabs/ASPA/commit/4b15013ea82b8a2cc9b749c04288d42928f6ae46

File: draft-ietf-sidrops-aspa-verification.xml (hunk 3 above)
[the remark this answers is not in the document]
Comment (ksriram25) [reply draft-ietf-sidrops-aspa-verification.xml:340]: @cjeker 

> Why the path is not valley-free is out of scope since ASPA is unable to identify the cause.

This suggestion is followed in the revised Sec. 7 that I just sent by email for comments.

File: draft-ietf-sidrops-aspa-verification.xml (hunk 5 above)
[the remark this answers is not in the document]
Comment (ksriram25) [reply draft-ietf-sidrops-aspa-verification.xml:199]: @cjeker 
I will make changes in Section 5 per your suggestions. I will provide much simplified/shorter text in Section 5. However, it is a bit hard to replace 'Provider+' and 'Not Provider+' with something else that is as intuitive. Because we need two 'Not Provider' hops (left and right) in opposite direction to detect Invalid. See the new Sec. 7 (Fig. 2) that I sent in email today. With 'Not Provider+' notation (connotes P2C or P2P) the explanation seems more intuitive for the reader. I will try to think a bit more here.

File: draft-ietf-sidrops-aspa-verification.xml (hunk 7 above)
[the remark this answers is not in the document]
Comment (ksriram25) [reply draft-ietf-sidrops-aspa-verification.xml:349]: >While better I would prefer something where upflow is not explained with the term up.

OK. I get it now. I will make the following substitution:
s/The term "upstream path" generally refers to AS paths received in the
   upstream direction (from a customer, lateral peer, RS-client, RS) and
   "downstream path" refers to AS paths received in the downstream
   direction (from a provider or mutual-transit neighbor)./
   The term "upstream path" refers to AS paths received 
   from a customer, lateral peer, RS-client, or RS, and
   "downstream path" refers to AS paths received
   from a provider or mutual-transit neighbor.

File: draft-ietf-sidrops-aspa-verification.xml (hunk 4 above)
[the remark this answers is not in the document]
Comment (ksriram25) [reply draft-ietf-sidrops-aspa-verification.xml:196]: @cjeker 
>In my opinion a VAP is analogous to VRP?

OK, agreed. we'll will eliminate VAP-SPAS from the doc. Just use VAP in its place. We'll mention that only validated ASPA records are used in the AS path verification. We'll also mention that if multiple ASPAs exist for a customer-as, then the VAP represents the union of the SPAS from those ASPAs.

[the remark this answers is not in the document]
Comment (cjeker) [reply draft-ietf-sidrops-aspa-verification.xml:196]: Great. I think you covered all cases in the above.

File: draft-ietf-sidrops-aspa-verification.xml (hunk 7 above)
[the remark this answers is not in the document]
Comment (cjeker) [reply draft-ietf-sidrops-aspa-verification.xml:349]: This is better. Will check in the new draft and comment there.

File: draft-ietf-sidrops-aspa-verification.xml (hunk 5 above)
[the remark this answers is not in the document]
Comment (cjeker) [reply draft-ietf-sidrops-aspa-verification.xml:199]: I see not that much of an issue with using different terms:
```
Because we need two 'not-in-set' hops (left and right) in opposite direction to detect Invalid.
```

Will look at the new branch and comment there.

File: draft-ietf-sidrops-aspa-verification.xml (hunk 1 above)
Comment (ksriram25) [reply draft-ietf-sidrops-aspa-verification.xml:669]: @cjeker @mitradir 
>Even OTC does not detect all route leaks initiated by the local AS (even if the provider AS has RFC9234 implemented) since it depends on the transitive nature of the OTC attribute. 

Claudio: That is not correct. Assume the local AS is doing OTC. If the OTC attribute is not present on a route received from a provider or lateral peer, then the local sets it on eBGP ingress and then the route will not be propagated to a provider or lateral peer on eBGP egress.

**Comment by ksriram25** [comment]
The next PR will include this PR #22 and more edits rebased on the Master.
